### PR TITLE
Add AMSMonitor interface and unify both RMQ API (#32)

### PR DESCRIPTION
**Pull request by lpottier**
`AMSMonitor` helps us monitor each stage of the python pipeline with simple decorator approach. Each call to `AMSMonitor` decorator or context manager populate a shared dictionary that track each call.

In addition of recording (by default, can be restricted with `@AMSMonitor(record=["attr1", "attr2"])`) all attributes of a class, AMSMonitor will also inject its own metrics, for example the duration of each block `amsmonitor_duration`. For example, the following code:

```python
class ExampleTask1(Task):
  def __init__(self):
    self.total_bytes = 0
  @AMSMonitor()
  def __call__(self):
    i = 0
    with AMSMonitor(obj=self, tag="while_loop"):
      while (i<3):
        self.total_bytes += 10
        time.sleep(1)
        i += 1
    self.total_bytes += 5
```
will produce the following dictionary (dumped as JSON here): 
```json
{
  "ExampleTask1": {
    "while_loop": {
      "02/29/2024-19:27:53": {
        "total_bytes": 30,
        "amsmonitor_duration": 3.004607439041138
      }
    },
    "__call__": {
      "02/29/2024-19:29:24": {
        "total_bytes": 35,
        "amsmonitor_duration": 3.10461138
      }
    }
  }
}
```

This PR also addresses #32 and unify both RMQ API.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -77,13 +77,18 @@ exclude = [
...
-
+exclude = [
+  ".git",
+  "__pycache__",
+  "*.egg-info",
+  "build"
+]
 # change the default line length number or characters.
 line-length = 120
+lint.select = ['E', 'F', 'W', 'A', 'PLC', 'PLE', 'PLW', 'I', 'N', 'Q']
```
Comment (koparasy) [inline pyproject.toml:90]: What does this do?

File: src/AMSWorkflow/ams/monitor.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+            # Example: we do not want to record total_bytes
+            # but just total_bytes2
+            #
+            # @AMSMonitor() would record all attributes
+            # (total_bytes and total_bytes2)
+            #
+            # @AMSMonitor(accumulate=True) would record all
+            #   attributes and accumulate their values
+            #   (sum up total_bytes instead of recording a
+            #   different total_bytes per invocation)
```
Comment (koparasy) [inline src/AMSWorkflow/ams/monitor.py]: Does this mean that by default monitor captures all attributes, and when you specify something it only captures the specified values? Can you make the comment clear?

**Review by koparasy — APPROVED** [review]
LGTM. I have some minor comments.

File: pyproject.toml (hunk 1 above)
Comment (lpottier) [reply pyproject.toml:90]: It activates a bunch of formatting rules (see https://docs.astral.sh/ruff/rules/) for `ruff`. I tried to use `flake8` but it's too slow so I moved to `ruff`. For example, `F` will activate all the `Pyflakes` rules.

File: src/AMSWorkflow/ams/monitor.py (hunk 2 above)
Comment (lpottier) [reply src/AMSWorkflow/ams/monitor.py]: I updated the comments, thanks.

**Pull request merged**